public.tf

```diff
@@ -0,0 +1,74 @@
+module "public_label" {
+  source     = "git::https://github.com/cloudposse/terraform-null-label.git?ref=tags/0.2.2"
+  namespace  = "${var.namespace}"
+  name       = "${var.availability_zone}"
+  attributes = ["public"]
+  stage      = "${var.stage}"
+  delimiter  = "${var.delimiter}"
+  tags       = "${var.tags}"
+}
+
+resource "aws_subnet" "public" {
+  count             = "${var.type == "public" ? length(var.names) : 0}"
+  vpc_id            = "${var.vpc_id}"
+  availability_zone = "${var.availability_zone}"
+  cidr_block        = "${cidrsubnet(var.cidr_block, ceil(log(length(var.names), 2)), count.index)}"
+
+  tags = {
+    "Name"      = "${module.public_label.id}${var.delimiter}${element(var.names, count.index)}"
+    "Stage"     = "${module.public_label.stage}"
+    "Namespace" = "${module.public_label.namespace}"
+  }
+}
+
+resource "aws_route_table" "public" {
+  count  = "${var.type == "public" ? length(var.names) : 0}"
+  vpc_id = "${var.vpc_id}"
+
+  tags = {
+    "Name"      = "${module.public_label.id}${var.delimiter}${element(var.names, count.index)}"
+    "Stage"     = "${module.public_label.stage}"
+    "Namespace" = "${module.public_label.namespace}"
+  }
+}
+
+resource "aws_route" "public" {
+  count                  = "${var.type == "public" ? length(var.names) : 0}"
+  route_table_id         = "${element(aws_route_table.public.*.id, count.index)}"
+  gateway_id             = "${var.igw_id}"
+  destination_cidr_block = "0.0.0.0/0"
+}
+
+resource "aws_route_table_association" "public" {
+  count          = "${var.type == "public" ? length(var.names) : 0}"
+  subnet_id      = "${element(aws_subnet.public.*.id, count.index)}"
+  route_table_id = "${element(aws_route_table.public.*.id, count.index)}"
+}
+
+resource "aws_network_acl" "public" {
+  count      = "${var.type == "public" && signum(length(var.private_network_acl_id)) == 0 ? 1 : 0}"
+  vpc_id     = "${data.aws_vpc.default.id}"
+  subnet_ids = ["${aws_subnet.public.*.id}"]
+  egress     = "${var.public_network_acl_egress}"
+  ingress    = "${var.public_network_acl_ingress}"
+  tags       = "${module.public_label.tags}"
+}
+
+resource "aws_eip" "default" {
+  count = "${var.type == "public" ? 1 : 0}"
+  vpc   = true
+
+  lifecycle {
+    create_before_destroy = true
+  }
+}
+
+resource "aws_nat_gateway" "default" {
+  count         = "${var.type == "public" ? 1 : 0}"
```

Comment:
creation of `NGW` should be parametrized for case if needs only subnets without `NGW`, I mean something like this:
```terraform
count         = "${var.type == "public" && nat_enabled == "true" ? 1 : 0}"
```

and set `nat_enabled ` to `true` by default

Reply:
Is there a use case for `nat_enabled`?

Reply:
following example will create 2 `NGW` in same AZ, but main idea was use one NGW for AZ

```terraform
module "public_subnets" {
  source            = "git::https://github.com/cloudposse/terraform-aws-named-subnets.git?ref=master"
  namespace         = "${var.namespace}"
  stage             = "${var.stage}"
  names             = ["apples", "oranges", "grapes"]
  vpc_id            = "${module.vpc.vpc_id}"
  cidr_block        = "10.0.2.0/24"
  type              = "public"
  igw_id            = "${module.vpc.igw_id}"
  availability_zone = "us-east-1a"
}

module "public_another_subnets" {
  source            = "git::https://github.com/cloudposse/terraform-aws-named-subnets.git?ref=master"
  namespace         = "${var.namespace}"
  stage             = "${var.stage}"
  names             = ["apples"]
  vpc_id            = "${module.vpc.vpc_id}"
  cidr_block        = "10.0.1.0/24"
  type              = "public"
  igw_id            = "${module.vpc.igw_id}"
  availability_zone = "us-east-1a"
}
```

Reply:
If we use 1 module instance, instead of 2, we won't face this issue 
```
module "public_subnets" {
  source            = "git::https://github.com/cloudposse/terraform-aws-named-subnets.git?ref=master"
  namespace         = "${var.namespace}"
  stage             = "${var.stage}"
  names             = ["apples", "oranges", "grapes", "apples2"]
  vpc_id            = "${module.vpc.vpc_id}"
  cidr_block        = "10.0.2.0/24"
  type              = "public"
  igw_id            = "${module.vpc.igw_id}"
  availability_zone = "us-east-1a"
}
```

Reply:
I mean, we may not need 2 module instances in the same AZ.

Reply:
everything can be :) , but one optional variable doesn't break the module or usability of it. you can looks on it like on additional feature

Reply:
@SweetOps Is there a use case of 2 module instances in a single AZ?

Reply:
in my example using two different CIDRs and another use case when just need to create subnets, so `NGW` will be useless

Reply:
lgtm, need any version of module for finishing current sprint